File: python/38.外观数列.py
#
# @lc app=leetcode.cn id=38 lang=python3
#
# [38] 外观数列
#


# @lc code=start
class Solution:
    # 官解
    def countAndSay(self, n: int) -> str:
        pre = '1'
        for i in range(1, n):
            start = 0
            post = 0
            cur = ''
            while post < len(pre):
                while post < len(pre) and pre[start] == pre[post]:
                    post += 1
                cur += str(post - start) + pre[start]
                start = post
            pre = cur
        return pre

    # 官解，直接枚举
    def countAndSay2(self, n: int) -> str:
        return '枚举打表'

    # 自写，可读性欠佳
    def countAndSay1(self, n: int) -> str:
        if n == 1:
            return '1'
        curIndex = 1
        curStr = '1'
        while curIndex < n:
            length = len(curStr)
            l = 0
            r = 0
            count = 0
            nextStr = ''
            while r < length:
                if curStr[l] == curStr[r]:
                    count += 1
                    r += 1
                else:
                    nextStr += str(count) + curStr[l]
                    count = 0
                    l = r
                if r == length:
                    nextStr += str(count) + curStr[l]
            curIndex += 1
            curStr = nextStr
        return curStr


# @lc code=end

File: cpp/38.外观数列.cpp
/*
 * @lc app=leetcode.cn id=38 lang=cpp
 *
 * [38] 外观数列
 */
#include <algorithm>
#include <iostream>
#include <string>
#include <unordered_map>
#include <unordered_set>
#include <vector>

using namespace std;
// @lc code=start
class Solution
{
public:
    // 遍历生成
    string countAndSay(int n)
    {
        string pre = "1";

        for (int i = 2; i <= n; i++) {
            string curr;
            /**
             * 定义了一个start和一个pos非常巧妙，通过两者位置的差值，计算出个数
             * 避免了使用其他的结构来存储值和频次。
            */
            int start = 0;
            int pos   = 0;
            while (pos < pre.length()) {
                /**
                 * 因为start和pos是相同的，所以启动的时候总是能计算出当前值的个数
                 * 看似其他地方没有给pos赋值，实际通过start = pos，让外层的循环条件
                 * while (pos < pre.length())能够结束。
                */
                while (pos < pre.length() && pre[pos] == pre[start]) {
                    pos++;
                }
                curr += to_string(pos - start) + pre[start];
                start = pos;
            }
            pre = curr;
        }

        return pre;
    }
};
// @lc code=end
